src/t_hash.c

```diff
@@ -32,233 +32,119 @@
-        size_t offset = sdslen(entry) + 1 + sdsHdrSize(SDS_TYPE_8);
-        return (char *)entry + offset;
-    }
-}
-
-/* Returns the address of the entry allocation. */
-static void *hashTypeEntryAllocPtr(hashTypeEntry *entry) {
-    char *buf = sdsAllocPtr(entry);
-    if (entryHasValuePtr(entry)) {
-        buf -= sizeof(sds *);
-    }
-    return buf;
-}
-
-/* Frees previous value, takes ownership of new value, returns entry (may be
- * reallocated). */
-static hashTypeEntry *hashTypeEntryReplaceValue(hashTypeEntry *entry, sds value) {
-    sds field = (sds)entry;
-    size_t field_size = sdsHdrSize(sdsType(field)) + sdsalloc(field) + 1;
-    size_t value_len = sdslen(value);
-    size_t value_size = sdsReqSize(value_len, SDS_TYPE_8);
-    if (!entryHasValuePtr(entry)) {
-        /* Reuse the allocation if the new value fits and leaves no more than
-         * 25% unused space after replacing the value. */
-        char *alloc_ptr = sdsAllocPtr(entry);
-        size_t required_size = field_size + value_size;
-        size_t alloc_size;
-        if (required_size <= EMBED_VALUE_MAX_ALLOC_SIZE &&
-            required_size <= (alloc_size = hashTypeEntryMemUsage(entry)) &&
-            required_size >= alloc_size * 3 / 4) {
-            /* It fits in the allocation and leaves max 25% unused space. */
-            sdswrite(alloc_ptr + field_size, alloc_size - field_size, SDS_TYPE_8, value, value_len);
-            sdsfree(value);
-            return entry;
-        }
-        hashTypeEntry *new_entry = hashTypeCreateEntry(hashTypeEntryGetField(entry), value);
-        freeHashTypeEntry(entry);
-        return new_entry;
-    } else {
-        /* The value pointer is located before the embedded field. */
-        if (field_size + value_size <= EMBED_VALUE_MAX_ALLOC_SIZE) {
-            /* Convert to entry with embedded value. */
-            hashTypeEntry *new_entry = hashTypeCreateEntry(field, value);
-            freeHashTypeEntry(entry);
-            return new_entry;
-        } else {
-            /* Not embedded value. */
-            sds *value_ref = hashTypeEntryGetValueRef(entry);
-            sdsfree(*value_ref);
-            *value_ref = value;
-            return entry;
+void hashTypeFreeVolatileSet(robj *o) {
+    volatile_set *set = hashTypeGetVolatileSet(o);
+    if (set)
+        freeVolatileSet(set);
+}
+
+bool hashTypeHasVolatileElements(robj *o) {
+    return ((o->encoding == OBJ_ENCODING_HASHTABLE) && (hashTypeGetVolatileSet(o) != NULL));
+}
+
+/* make any access to the hash object elements ignore the specific elements expiration.
+ * This is mainly in order to be able to access hash elements which are already expired. */
+static inline void hashTypeIgnoreTTL(robj *o, bool ignore) {
+    if (o->encoding == OBJ_ENCODING_HASHTABLE) {
+        /* prevent placing access function if not needed */
+        if (!ignore && !hashTypeHasVolatileElements(o)) {
+            ignore = 0;
```

Comment:
this looks wrong to me. should we set `ignore` to true instead?
```suggestion
            ignore = true;
```